### PR TITLE
Completely refactor error handling

### PR DESCRIPTION
**Pull request by ntruessel**

### CODE REVIEW

**Comment by jocelyn** [comment]
I don't see any caller of the various report errors project.
And not sure why you had to use only STRING_8 

Otherwise, yes this is an approach, using callback.
In various projects, I am using the error lib : https://svn.eiffel.com/eiffelstudio/trunk/Src/contrib/library/utility/general/error/

I realize now, it is missing a readme or doc, I will try to add that soon.
but tests/test_error.e shows how it could be used, especially synchronizing one or many error handlers.
Maybe one would need to implement a "one way synchronization" i.e just a "propagation".

Since the related issue #48 is closed, I am not sure you still want me to review this pull request.
I will wait for your reply.

**Comment by ntruessel** [comment]
I don't understand what you mean by "callers of the various report errors project", could you explain please? I left the existing interface mostly untouched, so errors are reported by `report_error ("Some description")`, `has_error` returns whether an error occurred in `Current` (or anyone it is listening to), and `error_messages` stores all reported errors.

I had to use `STRING_8` because of cat-calls (I do not fully understand why, because I thought `READABLE_STRING_GENERAL` would be fine).

Basically I wanted error propagation instead of synchronization, because e.g. the header parser should not care about errors that do not occur in his part of the code (failed setting metadata...). Furthermore, I think of adding warnings as well, so I prefer sticking to this code.

#48 was an outdated PR that was closed and replaced by this one.

**Comment by jocelyn** [comment]
> I don't understand what you mean by "callers of the various report errors
> project", could you explain please? I left the existing interface mostly
> untouched, so errors are reported by report_error ("Some description"),
> has_error returns whether an error occurred in Current (or anyone it is
> listening to), and error_messages stores all reported errors.

Ok since you did not changed the existing interfaces, that's why I was not
able to see any caller, the diff is just a partial view.
So forget about my comment.

I had to use STRING_8 because of cat-calls (I do not fully understand why,

> because I thought READABLE_STRING_GENERAL would be fine).
> 
> ok, I will try to find time to compile and play with it. Because I guess
> you may need unicode error message, maybe to report error related to a
> specific path (which may contain unicode character) .
> 
> Basically I wanted error propagation instead of synchronization, because
> e.g. the header parser should not care about errors that do not occur in
> his part of the code (failed setting metadata...). Furthermore, I think of
> adding warnings as well, so I prefer sticking to this code.

Ok, fine with me. I was just mentionning the error lib, in case it may
answer your needs.
On my side, I will see, if I can add one-way-sync, (i.e propagation) to the
current error lib.
